superset-frontend/src/SqlLab/components/QueryAutoRefresh/index.tsx

```diff
@@ -0,0 +1,118 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { useState, useEffect } from 'react';
+import { SupersetClient } from '@superset-ui/core';
+import * as Actions from 'src/SqlLab/actions/sqlLab';
+import { Query } from 'src/SqlLab/types';
+import useInterval from '../../utils/useInterval';
+
+const QUERY_UPDATE_FREQ = 2000;
+const QUERY_UPDATE_BUFFER_MS = 5000;
+const MAX_QUERY_AGE_TO_POLL = 21600000;
+const QUERY_TIMEOUT_LIMIT = 10000;
+
+interface UserOfflineFunc {
+  (offline: boolean): boolean;
+}
+
+interface RefreshQueriesFunc {
+  (alteredQueries: any): any;
+}
+
+interface Actions {
+  setUserOffline: UserOfflineFunc;
+  refreshQueries: RefreshQueriesFunc;
+}
+
+export interface QueryAutoRefreshProps {
+  offline: boolean;
+  queries: Query[];
+  actions: Actions;
+  queriesLastUpdate: number;
+}
+
+// returns true if the Query.state matches one of the specifc values indicating the query is still processing on server
+export const isQueryRunning = (q: Query): boolean =>
+  ['running', 'started', 'pending', 'fetching'].indexOf(q?.state) >= 0;
+
+// returns true if at least one query is running and within the max age to poll timeframe
+export const shouldCheckForQueries = (queryList: Query[]): boolean => {
+  let shouldCheck = false;
+  // if there are started or running queries, this method should return true
+  const now = Date.now();
+  if (queryList && typeof queryList === 'object') {
+    shouldCheck = Object.values(queryList).some(
+      q => isQueryRunning(q) && now - q?.startDttm < MAX_QUERY_AGE_TO_POLL,
+    );
+  }
+  return shouldCheck;
+};
+
+function QueryAutoRefresh({
+  offline,
+  queries,
+  actions,
+  queriesLastUpdate,
+}: QueryAutoRefreshProps) {
+  // In case of an error / timeout requesting query execution state
+  // we set isOffline to true.  the interval timer will continue to run 
+  // and update status if connection succeeds later
+  const [isOffline, setIsOffline] = useState(offline);
+  // We do not want to spam requests in the case of slow connections and potentially recieve responses out of order
+  // pendingRequest check ensures we only have one active http call to check for query statuses
+  const [pendingRequest, setPendingRequest] = useState(false);
+
+  useEffect(() => {
+    if (isOffline !== offline) {
+      actions?.setUserOffline(isOffline);
+    }
+  }, [isOffline, offline, actions]);
+
+  const refreshQueries = () => {
+    if (!pendingRequest && (shouldCheckForQueries(queries) || isOffline)) {
+      setPendingRequest(true);
+      SupersetClient.get({
+        endpoint: `/superset/queries/${
+          queriesLastUpdate - QUERY_UPDATE_BUFFER_MS
+        }`,
+        timeout: QUERY_TIMEOUT_LIMIT,
+      })
+        .then(({ json }) => {
+          if (json && Object.keys(json)?.length > 0) {
+            actions?.refreshQueries(json);
```

Comment:
is there ever a case where you want to clear out all of the queries, e.g., the json comes back blank?

Reply:
I was not sure why it was originally written this way so I had left it alone.  I removed the check and found no side affect / change to behavior and updated to remove the Object.keys check